File: internal/impl/telegram/message.go
// Copyright 2025 Redpanda Data, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package telegram

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/go-telegram/bot/models"
	"github.com/redpanda-data/benthos/v4/public/service"
)

// parseUpdate converts a Telegram Update into a Benthos message.
// The message content is the JSON-serialized Update object.
// Metadata includes chat_id, user_id, message_id, and timestamp for easy access.
func parseUpdate(update *models.Update) (*service.Message, error) {
	if update == nil {
		return nil, fmt.Errorf("nil update")
	}

	// Serialize the entire update as JSON
	content, err := json.Marshal(update)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize update: %w", err)
	}

	msg := service.NewMessage(content)

	// Extract and set metadata from the update
	extractMetadata(update, msg)

	return msg, nil
}

// extractMetadata extracts common fields from the Update and sets them as message metadata.
func extractMetadata(update *models.Update, msg *service.Message) {
	// Set the update ID
	msg.MetaSet("update_id", fmt.Sprintf("%d", update.ID))

	// Extract metadata based on update type
	var chatID int64
	var userID int64
	var messageID int
	var timestamp time.Time
	var messageType string

	switch {
	case update.Message != nil:
		messageType = "message"
		chatID = update.Message.Chat.ID
		if update.Message.From != nil {
			userID = update.Message.From.ID
		}
		messageID = update.Message.ID
		timestamp = time.Unix(update.Message.Date, 0)

	case update.EditedMessage != nil:
		messageType = "edited_message"
		chatID = update.EditedMessage.Chat.ID
		if update.EditedMessage.From != nil {
			userID = update.EditedMessage.From.ID
		}
		messageID = update.EditedMessage.ID
		timestamp = time.Unix(update.EditedMessage.Date, 0)

	case update.ChannelPost != nil:
		messageType = "channel_post"
		chatID = update.ChannelPost.Chat.ID
		if update.ChannelPost.From != nil {
			userID = update.ChannelPost.From.ID
		}
		messageID = update.ChannelPost.ID
		timestamp = time.Unix(update.ChannelPost.Date, 0)

	case update.EditedChannelPost != nil:
		messageType = "edited_channel_post"
		chatID = update.EditedChannelPost.Chat.ID
		if update.EditedChannelPost.From != nil {
			userID = update.EditedChannelPost.From.ID
		}
		messageID = update.EditedChannelPost.ID
		timestamp = time.Unix(update.EditedChannelPost.Date, 0)

	case update.CallbackQuery != nil:
		messageType = "callback_query"
		if update.CallbackQuery.Message != nil {
			if msg := update.CallbackQuery.Message.Message; msg != nil {
				chatID = msg.Chat.ID
				messageID = msg.ID
			}
		}
		userID = update.CallbackQuery.From.ID
		timestamp = time.Now()

	case update.InlineQuery != nil:
		messageType = "inline_query"
		userID = update.InlineQuery.From.ID
		timestamp = time.Now()

	default:
		messageType = "unknown"
		timestamp = time.Now()
	}

	// Set metadata
	if chatID != 0 {
		msg.MetaSet("chat_id", fmt.Sprintf("%d", chatID))
	}
	if userID != 0 {
		msg.MetaSet("user_id", fmt.Sprintf("%d", userID))
	}
	if messageID != 0 {
		msg.MetaSet("message_id", fmt.Sprintf("%d", messageID))
	}
	msg.MetaSet("message_type", messageType)
	msg.MetaSet("timestamp", timestamp.Format(time.RFC3339))
}

File: public/plugin/python/v1/src/redpanda_connect_v1/_message.py
# Copyright 2026 Redpanda Data, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Lightweight Message + MessageBatch wrappers that match the shape of
the benthos service.Message / service.MessageBatch types from a plugin
author's point of view."""

from __future__ import annotations

from typing import Any


class Message:
    """A single message flowing through a plugin component.

    The payload is either ``bytes`` or a structured Python value (dict,
    list, scalar). Metadata is a free-form dict. ``error`` is set by
    upstream components when a per-message failure should be propagated
    without aborting the whole batch.
    """

    __slots__ = ("_payload", "_metadata", "_error")

    def __init__(
        self,
        payload: bytes | str | dict[str, Any] | list[Any] | int | float | bool | None = None,
        metadata: dict[str, Any] | None = None,
        error: str | None = None,
    ) -> None:
        if isinstance(payload, str):
            self._payload: Any = payload.encode("utf-8")
        else:
            self._payload = payload
        self._metadata: dict[str, Any] = dict(metadata) if metadata else {}
        self._error = error

    # ------------------------------------------------------------------
    # payload accessors mirroring (*service.Message).AsBytes / .SetBytes
    # ------------------------------------------------------------------

    def as_bytes(self) -> bytes:
        """Return the payload as bytes. Raises ``TypeError`` if the
        payload is structured."""
        if isinstance(self._payload, bytes):
            return self._payload
        raise TypeError(f"payload is not bytes (got {type(self._payload).__name__})")

    def set_bytes(self, data: bytes | str) -> None:
        """Replace the payload with new bytes."""
        if isinstance(data, str):
            self._payload = data.encode("utf-8")
        else:
            self._payload = data

    def as_structured(self) -> Any:
        """Return the structured payload (dict/list/scalar)."""
        if isinstance(self._payload, bytes):
            raise TypeError("payload is bytes, not structured")
        return self._payload

    def set_structured(self, value: Any) -> None:
        """Replace the payload with a structured value."""
        self._payload = value

    # ------------------------------------------------------------------
    # metadata + error
    # ------------------------------------------------------------------

    @property
    def metadata(self) -> dict[str, Any]:
        return self._metadata

    @property
    def error(self) -> str | None:
        return self._error

    @error.setter
    def error(self, value: str | None) -> None:
        self._error = value

    @property
    def payload(self) -> Any:
        """Direct access to the raw payload (bytes or structured)."""
        return self._payload


# MessageBatch is just a list of Messages. Plugin authors iterate,
# mutate in place, and return a list-of-batches in process_batch.
MessageBatch = list[Message]
